DKGRituals/src/coordinator.ts

```diff
@@ -0,0 +1,126 @@
+    event.params.ritualId,
+  );
+
+  const ritual = new Ritual(event.params.ritualId.toString());
+  ritual.initiator = ritualInst.getInitiator();
+  ritual.initTimestamp = ritualInst.getInitTimestamp();
+  ritual.endTimestamp = ritualInst.getEndTimestamp();
+  ritual.authority = event.params.authority;
+  ritual.dkgSize = ritualInst.getDkgSize();
+  ritual.threshold = ritualInst.getThreshold();
+  ritual.accessController = ritualInst.getAccessController();
+  ritual.participants = changetype<Bytes[]>(event.params.participants);
+  ritual.postedTranscriptsAmount = 0;
+  ritual.postedTranscripts = [];
+  ritual.postedAggregationsAmount = 0;
+  ritual.postedAggregations = [];
+  ritual.dkgStatus = "DKG_AWAITING_TRANSCRIPTS";
+  ritual.save();
+
+  let ritualCounter = RitualCounter.load("Counter");
+  if (!ritualCounter) {
+    ritualCounter = new RitualCounter("Counter");
+    ritualCounter.total = 0;
+    ritualCounter.ended = 0;
+    ritualCounter.notEnded = 0;
+  }
+
+  ritualCounter.total = ritualCounter.total + 1;
+  ritualCounter.notEnded = ritualCounter.notEnded + 1;
+  ritualCounter.save();
+}
+
+export function handleTranscriptPosted(event: TranscriptPostedEvent): void {
+  const ritual = Ritual.load(event.params.ritualId.toString());
+  if (!ritual) {
+    log.error("Node {} posted a transcript for unknown ritual: {}", [
+      event.params.node.toHexString(),
+      event.params.ritualId.toString(),
+    ]);
+    return;
+  }
+  const postedTranscripts = ritual.postedTranscripts;
+  postedTranscripts.push(event.params.node);
+  ritual.postedTranscripts = postedTranscripts;
+  ritual.postedTranscriptsAmount = ritual.postedTranscriptsAmount + 1;
+  ritual.save();
+}
+
+export function handleStartAggregationRound(
+  event: StartAggregationRoundEvent,
+): void {
+  const ritual = Ritual.load(event.params.ritualId.toString());
+  if (!ritual) {
+    log.error("Received StartAggregationRound event for unknown ritual", []);
+    return;
+  }
+  ritual.dkgStatus = "DKG_AWAITING_AGGREGATIONS";
+  ritual.save();
+}
+
+export function handleAggregationPosted(event: AggregationPostedEvent): void {
+  const ritual = Ritual.load(event.params.ritualId.toString());
+  if (!ritual) {
+    log.error("Node {} posted an aggregation for unknown ritual: {}", [
+      event.params.node.toHexString(),
+      event.params.ritualId.toString(),
+    ]);
+    return;
+  }
+  const postedAggregations = ritual.postedAggregations;
+  postedAggregations.push(event.params.node);
+  ritual.postedAggregations = postedAggregations;
+  ritual.postedAggregationsAmount = ritual.postedAggregationsAmount + 1;
+  ritual.save();
+}
+
+export function handleEndRitual(event: EndRitualEvent): void {
+  const ritual = Ritual.load(event.params.ritualId.toString());
+  if (!ritual) {
+    log.error("Received EndRitual event for unknown ritual {}", [
+      event.params.ritualId.toString(),
+    ]);
+    return;
+  }
+  ritual.dkgStatus = "ENDED";
```

Comment:
Should you care about "successful" vs "not successful"?

Reply:
> Should you care about "successful" vs "not successful"?

We could add it. What name do you think could be proper for this status? SUCCESFULLY_ENDED and UNSUCCESFULLY_ENDED?

Reply:
It depends on how granular you want/need to be for using the subgraph. 

Do you care about whether the ritual is in progress or not? If not, what is the value until it completes? Perhaps, on the lower end seems like you need at least 3 values if doing a dkg state kind of thing:
- IN_PROGRESS (initial value)
- SUCCESSFUL
- UNSUCCESSFUL

If you want to be more granular, take a look at the `RitualState` enum, https://github.com/nucypher/nucypher-contracts/blob/main/contracts/contracts/coordination/Coordinator.sol#L48, and how those values are used for the `getRitualState` function, https://github.com/nucypher/nucypher-contracts/blob/main/contracts/contracts/coordination/Coordinator.sol#L146. Of course looking at the `getRitualState` it uses values on the ritual to determine the state. You can do the same if you like eg. `aggregationMismatch` being true means that the dkg was invalid.

Is the subgraph purely for the ritual ceremony, and once the ceremony is over (successful or not) you don't care after the fact i.e. do you care about a ritual expiring at a later date (EXPIRED state value)...

Of course, you can start simple and then increase granularity, it just depends.

Reply:
@derekpierre I worked on these changes in #9.

I splitted ENDED status into SUCCESSFUL and UNSUCCESFUL.

> It depends on how granular you want/need to be for using the subgraph.

I would like to be as granular as possible. But unfortunately, I don't think it is possible to add more states. Note that the subgraph data is only updated when a new Ethereum event is emitted. So, since there is not a "timeout" event, the subgraph can't know if a pending Ritual reached the timeout or not.